File: PyProject/custom/merge_sort.py
def merge_sort(arr):
    if len(arr) > 1:
        mid = len(arr) // 2
        L = arr[:mid]
        R = arr[mid:]
        merge_sort(L)
        merge_sort(R)
        arr.clear()

        while len(L) > 0 and len(R) > 0:
            if L[0] > R[0]:
                arr.append(R[0])
                R.pop(0)
            else:
                arr.append(L[0])
                L.pop(0)
        for j in L:
            arr.append(j)
        for j in R:
            arr.append(j)


if __name__ == '__main__':
    a = [1, 9, 3, 0, 4, 10, 5, 4, 6]
    merge_sort(a)
    print(a)

File: GoProject/lazada/sort/merge_sort.go
package main

import "fmt"

func merge_sort(nums []int) []int {
	if len(nums) <= 1 {
		return nums
	}
	lall := len(nums)
	mid := lall / 2
	left := merge_sort(nums[0:mid])
	right := merge_sort(nums[mid:])

	result := make([]int, 0)

	ll, lr := len(left), len(right)

	i, j := 0, 0

	for i < ll && j < lr {
		if left[i] < right[j] {
			result = append(result, left[i])
			i = i + 1
		} else {
			result = append(result, right[j])
			j = j + 1
		}
	}

	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

func main() {
	arr := []int{10, 7, 8, 9, 1, 5}
	result := merge_sort(arr)
	fmt.Println(result)
}